[None] [fix] improve kvcache allocation in PyTorch runtime

qixiang-99 opened the pull request:
# fix/improve kvcache allocation in PyTorch runtime

## Background
We found significant memory under-allocation when using VSWA KvCacheManager. One reason is that PyTorch runtime uses estimated 'max_tokens' to constrain KV cache allocation, while 'max_tokens' is a strange(and potentially ambiguous) concept in Variable Sliding Window Attention KV Cache.

## TL;DR
Previously, we used `max_tokens` to limit the KV cache memory usage, but that method didn't work for VSWA's KV cache. This PR refines the KV cache estimation logic (mainly in `KVCacheCreator`) by directly enforcing a memory size limit. 
Besides, this PR also contains these minor fixes:
1. Improve estimation phase to avoid potential OOM issues for long sequence with CUDA Graph.
2. Add optional `torch.cuda.empty_cache` pytest fixture to avoid potential OOM in CI
Big shout out to @jaedeok-nvidia, @ixlmar and @Funatiq for their suggestions and reviews!

## Description

Currently, the PyTorch path includes a [step](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/pyexecutor/py_executor_creator.py#L396) that estimates available free memory for the KV cache, calculates the corresponding 'max_tokens', and [updates](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/pyexecutor/_util.py#L267) the kv_cache_config accordingly.


However, the existing logic significantly limits memory allocation, as it assumes a homogeneous attention model ([nvbug](https://nvbugspro.nvidia.com/bug/5380887)). 

Additionally, the 'max_tokens' concept doesn't align well with VSWA scenarios, as its definition is "_The maximum number of tokens that should be stored in the KV cache_". In VSWA case, "tokens stored in KV cache" is a bit "coarse-grained", as the in-window tokens are "fully in KV cache", while the out-of-window tokens are "partially in KV cache" because of non-homogeneous attention.

## Proposed Solution 

An alternative approach could be to **directly compute the allocated memory size during the estimation step**.

### Advantages:
1. Aligns naturally with the block distribution function `KVCacheManagerCpp.calculate_max_num_blocks` ([link](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/pyexecutor/resource_manager.py#L642)), which already expects a memory size parameter.
3. Provides greater flexibility, particularly beneficial for future scenarios such as dynamic memory allocation in VSWA.

### Implementation Details

1. This PR introduces a new field, `mMaxGpuTotalBytes` (exposed as `max_gpu_total_bytes` in Python), in `KvCacheConfig`.  
2. Currently, `KvCacheCreator` is responsible for building and estimating KV caches. With this PR, it also calculates the estimated in-memory size of the KV cache and uses the `max_gpu_total_bytes` setting from `KvCacheConfig` to limit memory allocation.  
4. Within `KVCacheManager`, the VSWA path uses a C++ binding function (see `calculate_max_num_blocks_from_cpp`) that takes a memory size as input to determine KV blocks. For non-VSWA cases, it still uses its own Python function, but future efforts aim to unify both approaches, as discussed [here](https://nvidia.slack.com/archives/C059LSY62BT/p1748549582863269).

### ~~"Hack" to achieve this~~
~~One hack is to interpret `freeGpuMemoryFraction` as **the fraction of total GPU memory** reserved for the KV cache. This allows us to avoid adding an extra field, but we should clearly document this usage to prevent confusion. (Perhaps add a clarifying comment in the code?)~~
 
## Test Coverage
`tests/unittest/_torch/test_resource_manager.py`
<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Add option to cap KV-cache GPU memory via max_gpu_total_bytes (used with existing fraction limit).
  - KV-cache stats now report allocated GPU bytes; exposed in Python.
  - Automatic KV-cache sizing moved to a memory-based approach, accounting for main/draft models and VSWA.

- **Refactor**
  - Public API renamed: estimate_max_tokens → configure_kv_cache_capacity; executor creation uses the new flow.

- **Tests**
  - New fixture clears CUDA cache before tests; applied to an existing disaggregated single-GPU test.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

qixiang-99 committed:
fix: WAR for under-allocation in torch VSWA kvcachemanager

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

feat: Add max_free_gpu_memory_size support for KV cache configuration

- Introduced max_free_gpu_memory_size to manage GPU memory allocation for KV cache.
- Updated KvCacheConfig and related methods to handle the new parameter.
- Modified estimation logic in KvCacheCreator to utilize max_free_gpu_memory_size for VSWA cases.
- Adjusted resource management to ensure compatibility with the new memory allocation strategy.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

address comments:
- vswa path should work now
- modify `KvCacheCreator` so it will always provide KvCacheManager memory size instead of max_tokens
- Handle user provided `max_tokens` inside `KvCacheCreator` -- it will raise warning for VSWA case; it will translate `maxt_tokens` to memory size for non-VSWA case

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

Respect max_gpu_total_bytes in KVCacheManager for non-VSWA case

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

add memory estimation for attention metadata to solve OOM issue when cuda graph is enabled and max window size is large

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

Enhance KVCacheManager to maintain adjusted max attention window sizes. Introduced an adjusted dictionary to track window size mappings and updated the logic to reflect these changes in the max attention window vector. Updated unit tests to validate the new behavior and ensure expected outputs for various memory configurations.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

fix: Draft model calculation in KVCacheManager and KvCacheCreator
- Introduced a method to calculate KV size per token and updated memory estimation to ensure proper handling of max_gpu_total_bytes.
- Changed KVCacheManager to still use `max_tokens` in non-VSWA case as its`calculate_max_num_blocks` doesn't consider the draft model tokens calculation.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

minor fix for description

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

Enhance KVCacheManager and related components to track allocated GPU memory for KV-cache. Added `allocatedBytes` to `KvCacheStats` and updated memory allocation logic in `KVCacheManager` and `KvCacheCreator` to ensure accurate memory usage reporting. Adjusted Python bindings to expose the new `allocated_bytes` attribute.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

update logging

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

update Nanobind accordingly

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

Update secondary pool memory allocation in KVCacheManager to use host_cache_size from configuration if provided.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

Refactor the fix for CUDA Graph OOM issue.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

Add torch_empty_cache fixture to clear CUDA cache before tests

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

minor fix for rebase

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

Add validation for max_gpu_total_bytes in KvCacheConfig

- Introduced a field validator to ensure that max_gpu_total_bytes is non-negative, enhancing input validation for the configuration.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #11576](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11576/) [ run ] triggered by Bot

jaedeok-nvidia commented on cpp/include/tensorrt_llm/executor/executor.h:
```diff
@@ -1085,6 +1088,12 @@ class KvCacheConfig
+    /// This is only used for VSWA case for now as a alternative to mMaxTokens.
+    /// If both mMaxFreeGpuMemorySize and mFreeGpuMemoryFraction are specified, memory corresponding to the minimum will
+    /// be allocated.
+    std::optional<uint64_t> mMaxFreeGpuMemorySize;
```
I think `mMaxFreeGpuMemorySize` is a redundant field, as we can calculate it from `fraction` ([here](https://github.com/NVIDIA/TensorRT-LLM/blob/8628139625f608b708857384d5952dbcb30179f4/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp#L2145-L2178)). If it's necessary, why don't we calculate at the torch backend side and pass `freeGpuMemoryFraction` instead? That allows us to keep the fields minimal.

jaedeok-nvidia replied:
Thanks! As we discussed offline and below thread, it looks ok for me now.

jaedeok-nvidia commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -94,8 +94,13 @@ def _get_free_gpu_memory_fraction(self) -> float:
+                                   alloc_kv_tokens: int) -> tuple[int, int]:
+        """
+        Calculate the max KV cache capacity in max_tokens and max_free_gpu_memory_size.
+        For VSWA case, we use max_free_gpu_memory_size instead of max_tokens.
```
Can I know why we need memory as well? Since, in VSWA case, we will set `max_tokens=None` before calculating the number of cache blocks, so `available_kv_mem` seems unnecessary.

ixlmar replied:
I'd propose to keep this simple: Make this `_calculate_available_memory(...) -> int` and let it return an upper limit on the total memory footprint (in bytes) of the draft+model KV caches.

jaedeok-nvidia commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -115,7 +120,7 @@ def _cal_max_tokens(self, peak_memory, total_gpu_memory, fraction,
         )
```
How about not printing `tmp kv_mem` field in case of VSWA not to make a confusion? It depends on maxTokens as well as adjusted window sizes which will be done later.

tensorrt-cicd commented:
[PR_Github #11576](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11576/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8571](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8571/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/8571/test-results/rerun_report.html) for details.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -185,8 +190,9 @@ def try_prepare_estimation(self) -> bool:
         return estimating_kv_cache
 
-    def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
+    def estimate_max_tokens_or_memory(self, py_executor: PyExecutor) -> None:
```
Is there a need to make VSWA a special case here? I think it should be sufficient to always `estimate_max_bytes(...) -> int`

ixlmar replied:
This implies that `KvCacheConfig.max_tokens` should be removed completely in favor of `max_total_bytes`.

jaedeok-nvidia replied:
Could you please help me understand why we need `max_tokens` or `max_total_bytes` here? Is this related to the draft model? IIUC, the process is as follows: KVCacheCreator
  1. sends a dummy request with the maximum sequence length,
  2. gathers the peak memory and cache usage, and
  3. sets the estimated max_tokens in the kv_cache_config.
But, even if we do not explicitly set `max_tokens` on the PyTorch side, the C++ KvCacheManager will calculate the maximum available tokens at [calculateMaxNumBlocks](https://github.com/NVIDIA/TensorRT-LLM/blob/6d4b045d1f9196e3ea2b0b4a422ac5fb512165ce/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp#L2210). 

Regarding `max_total_bytes`, it seems flexible, however, I am not sure if it will be useful for others since it cannot be directly used for calculating number of blocks.

One additional point is that for VSWA, the KV cache pools are currently separated per window. Therefore, if we want to estimate the KV cache size, we need to estimate the cache size per window.

qixiang-99 replied:
Hi @jaedeok-nvidia, sure, it would be great to have some clarifications to ensure we're aligned. Let me summarize my current understanding below:

Regarding the `KVCacheCreator`:
1. It sends dummy request(s) with a total token count of `min(maximum sequence length, max_num_tokens)`. My understanding is that this step determines the KV cache capacity required to successfully run **one iteration**, preventing OOM issues during subsequent dummy runs.
2. It doesn't directly estimate the KV cache required by a model, but rather estimates how much memory the model can utilize for the KV cache ([reference](https://github.com/NVIDIA/TensorRT-LLM/blob/6d4b045d1f9196e3ea2b0b4a422ac5fb512165ce/tensorrt_llm/_torch/pyexecutor/_util.py#L108-L109)).
3. I'm not entirely sure why we need dummy forward steps for KV cache estimation, but my guess is due to the difficulty in precisely calculating and controlling memory usage when running inference with PyTorch.

Regarding [`calculateMaxNumBlocks`](https://github.com/NVIDIA/TensorRT-LLM/blob/6d4b045d1f9196e3ea2b0b4a422ac5fb512165ce/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp#L2210):
1. My understanding is this function prefers memory size over token count (`max_tokens`), as indicated by [this warning](https://github.com/NVIDIA/TensorRT-LLM/blob/6d4b045d1f9196e3ea2b0b4a422ac5fb512165ce/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp#L2220-L2227). With the memory size, the C++ KvCacheManager will first split the memory size across different windows, and "_calculate the maximum available tokens (for each window size) at [calculateMaxNumBlocks](https://github.com/NVIDIA/TensorRT-LLM/blob/6d4b045d1f9196e3ea2b0b4a422ac5fb512165ce/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp#L2210)_"

Funatiq replied:
Regarding `free_gpu_memory_fraction`, `max_tokens`, `max_total_bytes`, they all make sense to me from a user perspective.
* `free_gpu_memory_fraction` - user doesn't know or care how much memory in bytes are available, but wants to have some control how much memory should be left free.
* `max_tokens` - user knows/estimates how many requests/tokens are processed at a time and wants to reserve specific token budget. To me this only makes sense if it considers the full request length, so either full attention (non-SWA) or VSWA with full attention layers (also see the warning [here](https://github.com/NVIDIA/TensorRT-LLM/blob/6d4b045d1f9196e3ea2b0b4a422ac5fb512165ce/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp#L2220)). We could simply disallow this option if no full attention layer is used.
* `max_total_bytes` - user wants to reserve a specific amount of memory, but doesn't care how many requests can run.

Therefore I don't think we should remove one in favor of another. What we could do, however, is restricting the `KvCacheConfig` so that we don't have to deal with all possible combinations of the options. E.g. restrict to either use `max_tokens` or `max_total_bytes`, or have a function that calculates the resulting bytes from `max_tokens` and take the minimum with `max_total_bytes` (and `free_gpu_memory_fraction`). Then we could always call `calculateMaxNumBlocks` with a single bytes value. WDYT?

jaedeok-nvidia replied:
Thank @qixiang-99 for kind explanation and thank @Funatiq / @ixlmar for discussions.

I agree both @Funatiq / @ixlmar. 

Currently, KvCacheManager splits cache pool per window, so that in practice VSWA model uses inefficient cache pool. For instance, for VSWA model having windows [1k, 100k], then KvCacheManager allocates pools proportional to window size. If we have 1) ISL 1k x 100 samples vs 2) ISL 100k x 1 samples. Number of ISL tokens are the same, however, case 1) will consume much more memory. And if cache pool for window=1k is no enough, then batch size will be limited because of unused pool of window=100k. So, unifying cache pool will be necessary. This is out-of-topic here, but someday we need to do.

Considering these, for now, I hoped we can have a minimal change (by minimal effort) at this moment.

And, after the above discussions, each parameter seems meaningful, and the current implementation looks good to me (I didn't want to against).

ixlmar replied:
@qixiang-99 Understood. To reduce technical debt given the direction suggested in https://github.com/NVIDIA/TensorRT-LLM/pull/5933#discussion_r2210770747, I propose to make the following hopefully quick changes to `KvCacheCreator`:

* Rename `estimate_max_tokens_or_memory` to `configure_kv_cache_capacity`.

* Change `_cal_max_tokens_and_memory` to `_cal_max_memory`. Handle everything except `max_tokens` there as done currently.

* In `configure_kv_cache_capacity`:

  * If user provided `max_tokens`,  use `_get_cache_size_per_token` to convert to bytes and update max. memory obtained via `_cal_max_memory` accordingly.
  * If VSWA, raise warning or error as done now. Since we are validating user-inputs we should probably `raise` a meaningful error rather than `assert`.

  * Only pass max. memory to C++ bindings in all cases.

qixiang-99 replied:
Thanks for the discussion, team! I've done these changes thanks to @ixlmar 's detailed suggestion:
1. Rename `estimate_max_tokens_or_memory` to `configure_kv_cache_capacity` ([here](https://github.com/NVIDIA/TensorRT-LLM/pull/5933/files#diff-5f0e349b724e8ad501e0ae859c992801e631217a8d52acfa40e24a8fe95bbbcbR248))
2. Change `_cal_max_tokens_and_memory` to `_cal_max_memory` and it will only return memory size in bytes.
3. Added function to handle the conversion from user-defined `max_tokens` to memory in bytes, and will raise Warning if it is VSWA model ([here](https://github.com/NVIDIA/TensorRT-LLM/pull/5933/files#diff-5f0e349b724e8ad501e0ae859c992801e631217a8d52acfa40e24a8fe95bbbcbR345-R353))

jaedeok-nvidia replied:
Thanks for catching this point. 

One more concern was, a dummy request could be too short to fill the sliding window size. When testing EXAONE4 model, I found dummy req has a seq of length 2080, which is much shorther than the window size 4096. That means estimating via `kv_stats` can be incorrect. 

I feel we may need to redesign related parts 😂 . Let's discuss it offline later.

ixlmar commented on tensorrt_llm/llmapi/llm_args.py:
```diff
@@ -796,6 +796,11 @@ class KvCacheConfig(BaseModel, PybindMirror):
     )
     use_uvm: bool = Field(default=False,
                           description="Whether to use UVM for the KV cache.")
+    max_free_gpu_memory_size: Optional[int] = Field(
```
I think `kv_cache_config.max_total_bytes` would be a more fitting name for this. Mainly, I do not think that the name should include "free".

jaedeok-nvidia replied:
Accordingly, it may need to be changed C++ part - `maxFreeGpuMemorySize` to `max_total_bytes`.

jaedeok-nvidia commented on cpp/include/tensorrt_llm/executor/executor.h:
```diff
@@ -1030,6 +1032,7 @@ class KvCacheConfig
     void setSecondaryOffloadMinPriority(std::optional<RetentionPriority> secondaryOffloadMinPriority);
     void setEventBufferMaxSize(size_t eventBufferMaxSize);
     void setUseUvm(bool useUvm);
+    void setMaxFreeGpuMemorySize(uint64_t maxFreeGpuMemorySize);
```
Should we allow `std::optional<unin64_t> maxFreeGpuMemorySize`? none means unset the limit.

qixiang-99 replied:
Thanks for the suggestion. I've removed the optional and rename the variable as `maxGpuTotalBytes`

qixiang-99 commented:
/bot run --disable-fail-fast

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #12399](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12399/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12399](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12399/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9218](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9218/) completed with status: 'FAILURE'

jaedeok-nvidia reviewed (approved):
Thank @qixiang-99 for your contribution to VSWA fixes. Hope this PR fix our underutilization issue in VSWA models. And, let's consider how to improve the efficient further.

tensorrt-cicd commented:
[PR_Github #12473](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12473/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9278](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9278/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #12509](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12509/) [ run ] triggered by Bot

qixiang-99 commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -255,16 +261,29 @@ def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
 
-        logger.info(f"Estimated max tokens in KV cache : {kv_cache_max_tokens}")
-        executor_config.kv_cache_config.max_tokens = kv_cache_max_tokens
+        if executor_config.kv_cache_config.max_attention_window is not None:
```
Updated, thanks!

jaedeok-nvidia commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -187,8 +245,9 @@ def try_prepare_estimation(self) -> bool:
         return estimating_kv_cache
 
-    def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
+    def configure_kv_cache_capacity(self, py_executor: PyExecutor) -> None:
```
Love this new function name. It makes more sense than before, as this function sets the cache capacity for the class.

qixiang-99 commented:
/bot run --disable-fail-fast

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #12770](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12770/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12770](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12770/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #9510](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9510/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #12882](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12882/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12881](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12881/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #12883](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12883/) [ run ] triggered by Bot

qixiang-99 commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -254,19 +313,72 @@ def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
+        # calculate max memory from peak memory and free gpu memory fraction
+        kv_cache_max_memory = self._cal_max_memory(
             peak_memory, total_gpu_memory, fraction,
             kv_stats.max_num_blocks * kv_stats.tokens_per_block)
```
This is a great observation! After reviewing `kv_stats.max_num_blocks` in the VSWA context, it does indeed sum `max_num_blocks` for each window size manager. Fundamentally, `_cal_max_memory` needs the total allocated memory for the KV cache during estimation, which counts toward available KV memory in a real run. We already calculate this allocated KV memory size in existing code (see [the relevant code section](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/batch_manager/kvCacheManager.cpp#L1733-L1759)), so I've exposed this value through `kv_stat`. This approach should be able to addresses the issue from "first principles" without having to manage a more complex calculation : )

tensorrt-cicd commented:
[PR_Github #12883](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12883/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9602](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9602/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #12897](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12897/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12897](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12897/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9615](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9615/) completed with status: 'SUCCESS'

ixlmar commented on cpp/tensorrt_llm/pybind/executor/executorConfig.cpp:
```diff
@@ -138,6 +138,8 @@ void initConfigBindings(pybind11::module_& m)
             "sink_token_length", &tle::KvCacheConfig::getSinkTokenLength, &tle::KvCacheConfig::setSinkTokenLength)
         .def_property("free_gpu_memory_fraction", &tle::KvCacheConfig::getFreeGpuMemoryFraction,
             &tle::KvCacheConfig::setFreeGpuMemoryFraction)
+        .def_property(
```
The Nanobind bindings need to be updated accordingly.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -97,8 +97,60 @@ def _get_free_gpu_memory_fraction(self) -> float:
+
+        • CUDA graphs: when graph capture is enabled the full
+          `kv_cache_block_offsets` tensor must be pre-allocated,
+          making the extra memory grow linearly with the number of graphs.
```
Can the memory for this tensor be shared between CUDA graphs?

qixiang-99 replied:
This is a good point, and I did look into it. As I'm not a CUDA Graph expert, please take this with a grain of salt: my understanding is that we really only need one `kv_cache_block_offsets` that's large enough for all block offset tracking; **from that perspective, it could likely be shared.**

However, implementing this would be non-trivial because the current hierarchy is: `PyTorchModelEngine` owning multiple `DecodingCUDAGraphRunner` instances for different batch sizes, each with its own `AttentionMetadata` (where `kv_cache_block_offsets` resides). 

I recommend addressing this in a separate pull request to keep this PR focused, WDYT?

ixlmar replied:
Yes, makes sense.

ixlmar commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -255,19 +315,80 @@ def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
+            kv_cache_max_memory = min(kv_cache_max_memory,
+                                      kv_cache_max_memory_from_max_tokens)
+            logger.info(
+                f"max_tokens is provided, max_memory is set to {kv_cache_max_memory / (GB):.2f} GiB"
```
Perhaps include `self._max_kv_tokens_in` in the log message.

qixiang-99 replied:
Updated, thanks!

ixlmar commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -255,19 +315,80 @@ def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
+                kv_cache_max_memory,
+                executor_config.kv_cache_config.max_gpu_total_bytes)
+            logger.info(
+                f"max_gpu_total_bytes is provided, max_memory is set to {kv_cache_max_memory / (GB):.2f} GiB"
```
Print `executor_config.kv_cache_config.max_gpu_total_bytes` as well.

tensorrt-cicd commented:
[PR_Github #13041](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13041/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13041](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13041/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9742](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9742/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #13089](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13089/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13089](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13089/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9787](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9787/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #13144](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13144/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13144](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13144/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9839](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9839/) completed with status: 'SUCCESS'

ixlmar reviewed (commented):
Reviewed latest changes

ixlmar reviewed (approved):
LGTM

Note that https://github.com/NVIDIA/TensorRT-LLM/blob/e9be1e9de4836f86e6fa36080270ff236b47a06e/tensorrt_llm/_torch/pyexecutor/_util.py#L244-L247 might currently be over-dimensioning the initial KV cache because it does not account for VSWA. OTOH, if it can be left as it is, this reduces code complexity.

QiJune commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -188,8 +247,9 @@ def try_prepare_estimation(self) -> bool:
         return estimating_kv_cache
 
-    def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
+    def configure_kv_cache_capacity(self, py_executor: PyExecutor) -> None:
```
@HuiGao-NV  Could you please help review this change? Thanks

HuiGao-NV commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -97,8 +97,60 @@ def _get_free_gpu_memory_fraction(self) -> float:
+        num_pools = kv_cache_manager.num_pools
+
+        # calculate the max_blocks_per_seq in real inference phase
+        real_phase_max_blocks_per_seq = int(
```
I suggest to make the calculation of max_blocks_per_seq be moved into modelEngine or inside of TrtllmAttentionMetadata. This shall be one of their attribute and responsibility.
Then, we can get the value at multiple places such as here and warmup stage to pre allocate all  max_blocks_per_seq.

HuiGao-NV commented on cpp/tensorrt_llm/executor/kvCacheConfig.cpp:
```diff
@@ -36,6 +36,7 @@ KvCacheConfig::KvCacheConfig(bool enableBlockReuse, std::optional<SizeType32> co
     , mEnablePartialReuse{enablePartialReuse}
     , mCopyOnPartialReuse{copyOnPartialReuse}
     , mUseUvm{useUvm}
+    , mMaxGpuTotalBytes{maxGpuTotalBytes}
```
I didn't see how this mMaxGpuTotalBytes impact on kvCacheManager or any logic else. Why do we need it?

HuiGao-NV commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -255,19 +315,80 @@ def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
+        if executor_config.kv_cache_config.max_gpu_total_bytes > 0:
+            kv_cache_max_memory = min(
+                kv_cache_max_memory,
+                executor_config.kv_cache_config.max_gpu_total_bytes)
```
I see what max_gpu_total_bytes means and what it does. I suggest to change its name to max_kv_bytes.

qixiang-99 commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -97,8 +97,60 @@ def _get_free_gpu_memory_fraction(self) -> float:
+          orders of magnitude larger, so the tensor consumes significantly more
+          GPU memory.
+
+        • CUDA graphs: when graph capture is enabled the full
```
Hi @HuiGao-NV, thanks for reviewing. Let me put things together here to make sure we are all on the same page:
1. Regarding the memory usage of `kv_cache_block_offsets`(shape=`[num_pools, max_num_sequences, 2, max_blocks_per_seq]`): `PyTorchModelEngine` owning multiple `DecodingCUDAGraphRunner` instances for different batch sizes, each with its own `AttentionMetadata` (where `kv_cache_block_offsets` resides).
2. When I refer to "Kv cache estimation phase", I mean these [codes](https://github.com/qixiang-99/TensorRT-LLM/blob/e61c6862e6f48c179ec24fc5b0f4cf38339ecb15/tensorrt_llm/_torch/pyexecutor/_util.py#L264-L317). And yes, you are right, the memory should be reflected in the Peak memory usage stats. However, **in Kv cache estimation phase**, we only set a small KV Cache(set max_tokens [here](https://github.com/qixiang-99/TensorRT-LLM/blob/e61c6862e6f48c179ec24fc5b0f4cf38339ecb15/tensorrt_llm/_torch/pyexecutor/_util.py#L237-L248), e.g. usually 8192) to run a small dummy request(e.g. one req with input_seq_len=8192). So the `max_blocks_per_seq` is small(e.g. `8192/32(tokens_per_block)=256`). So **the Peak memory usage only reflects the `kv_cache_block_offsets`, for exmaple, with shape=`[num_pools, max_num_sequences, 2, 256]`.**
3. When I refer to "Real inference", I mean the real py_executor creation in these [codes](https://github.com/qixiang-99/TensorRT-LLM/blob/636265a35b5aa23a2e3e51a8f8188e90ad196c77/tensorrt_llm/_torch/pyexecutor/py_executor_creator.py#L399-L436). In this phase, we will re-create KvCacheManager for the real inference([here](https://github.com/qixiang-99/TensorRT-LLM/blob/636265a35b5aa23a2e3e51a8f8188e90ad196c77/tensorrt_llm/_torch/pyexecutor/py_executor_creator.py#L408)). In real inference, the `max_blocks_per_seq` could be large for long sequence model--for example, for max_seq_len=128k, the `max_blocks_per_seq` could be `128k/32(tokens_per_block)=4096`. So `kv_cache_block_offsets.shape=[num_pools, max_num_sequences, 2, 4096]`. 
4. So this function is used to **calculate the difference in size between `[num_pools, max_num_sequences, 2, 256]`(estimation phase) and `[num_pools, max_num_sequences, 2, 4096]`** for each `kv_cache_block_offsets` , and the final result will **multiply the number of graphs**.
5. **In most cases we wouldn't hit any issue with this difference**, because we usually have some free memory and the diff wouldn't be too large.  I found this issue because in this [nvbug](https://nvbugspro.nvidia.com/bug/5380887), the reporter used 1310972 as the max_seq_len, which make the real run has 162x larger `kv_cache_block_offsets` than the estimation phase's one, and in total 162x34(num of graphs)=**5505x** larger.

TBH I am not an expert about the CUDA Graph and ModelEngine part, so please let me know if above explanation is missing something or incorrect. Thank you!

HuiGao-NV reviewed (approved):
LGTM

HuiGao-NV reviewed (commented):
After discussion with members, the problem may be max_seq_len is overwritten by max_tokens if max_tokens is smaller.
If so, the better place is in function try_prepare_estimation in file _util.py, change  block
```
self._executor_config.kv_cache_config.max_tokens = self._get_token_num_for_estimation(
            ) 
```
to 
```
max_attention_window_from_config = self._executor_config.kv_cache_config.max_attention_window
        max_window_size = max(
            max_attention_window_from_config
        ) if max_attention_window_from_config is not None else self._executor_config.max_seq_len
self._executor_config.kv_cache_config.max_tokens = self._executor_config.kv_cache_config.max_attention_window
```

Then, KV cache will take correct max_seq_len during memory estimation.

qixiang-99 commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -255,19 +315,80 @@ def estimate_max_tokens(self, py_executor: PyExecutor) -> None:
+                                                   allocated_bytes)
+
+        max_attention_window = executor_config.kv_cache_config.max_attention_window
+        is_vswa = max_attention_window and len(max_attention_window) > 1
```
Yes, [128, 128]  wouldn't trigger error here. I have confusion here, what does [128, 128] stand for?

qixiang-99 replied:
Previously I assumed the user will provide `max_attention_window` , now I see your point here, thanks! Will update to:
```python
is_vswa = max_attention_window and len(set(max_attention_window)) > 1
```
Sounds good?

tensorrt-cicd commented:
[PR_Github #13855](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13855/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13855](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13855/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10423](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10423/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #13891](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13891/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #13891](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/13891/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10456](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10456/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14030](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14030/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10579](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10579/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #14054](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14054/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14054](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14054/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10602](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10602/) completed with status: 'SUCCESS'

tensorrt-cicd commented:
[PR_Github #14239](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14239/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14243](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14243/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14339](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14339/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14339](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14339/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10836](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10836/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14353](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14353/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14353](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14353/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #10846](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/10846/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/10846/test-results/rerun_report.html) for details.

HuiGao-NV commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #14434](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14434/) [ reuse-pipeline ] triggered by Bot

HuiGao-NV commented:
/bot reuse-pipeline

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14562](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14562/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14566](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14566/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14682](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14682/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11083](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11083/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14693](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14693/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14717](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14717/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11108](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11108/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14718](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14718/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14718](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14718/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11109](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11109/) completed with status: 'FAILURE'

nv-guomingz reviewed (approved):
LGTM for LLM API part

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #14732](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14732/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11119](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11119/) completed with status: 'FAILURE'

eopXD commented on tests/unittest/_torch/test_resource_manager.py:
```diff
@@ -527,6 +534,13 @@ def test_adjust_window_sizes_for_vswa(self):
+                    adjusted_max_attention_window_vec,
+                    expected_max_attention_window_vec,
+                    f"Test case '{description}' failed.\n"
+                    f"Memory bytes: {memory_bytes}\n"
```
We also need test coverage on memory fraction calculated through `calculate_max_num_blocks_from_cpp`

jaedeok-nvidia replied:
Thanks @eopXD for the suggestion. Improving test coverage is important. However, if it's not super critical, let's add the test in a separate PR, because this PR has already been delayed for too long due to CI infra failures, while it's necessary to fix the vswa memory underutilization issue. If we add a new commit now, it might take a few more days 😢

QiJune replied:
There is a duplicated PR that address memory fraction part (https://github.com/NVIDIA/TensorRT-LLM/pull/6807)  so we will add the test at there.

qixiang-99 committed:
feat: Enhance max_attention_window validation in KvCacheConfig

- Added a field validator for max_attention_window to ensure it is a non-empty list of positive integers.
- Implemented checks to prevent redundant patterns in the list, ensuring only minimal repeating patterns are accepted.
- Updated KvCacheCreator to use a set for determining VSWA status based on max_attention_window.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
test: Add test to protect the memory allocation behavior.

- Changed primary and secondary pool memory allocation to use instance variables instead of local variables for better clarity and maintainability.
- Updated logging to reflect the new instance variable usage.
- Added unit tests to validate memory allocation behavior in KVCacheManager.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

jaedeok-nvidia commented on tensorrt_llm/llmapi/llm_args.py:
```diff
@@ -1036,6 +1036,37 @@ def validate_max_gpu_total_bytes(cls, v: int):
+                    "kv_cache_config.max_attention_window values must be positive"
+                )
+
+        # Must not be a redundant repetition of a shorter pattern
```
Should we prevent this case? Of course, most models do not have such case, but I'm not sure we have to prevent it from the argument.

qixiang-99 replied:
Thanks for the thoughtful review. I originally added this to simplify inputs by enforcing the shortest pattern. After reconsidering, I agree with you—the validation is too strict because:

- The description doesn’t require user to do it.
- Some HF model configs don’t use the shortest pattern.

I’ll remove this validation. Thanks for flagging it!

jaedeok-nvidia replied:
Thanks for the quick update 😄

jaedeok-nvidia commented:
@qixiang-99 thanks for addressing the comments and test to improve test coverage. I left a question, but overall it looks good to me. Thanks!

tensorrt-cicd commented:
[PR_Github #15075](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15075/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11383](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11383/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #15190](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15190/) [ run ] triggered by Bot

jaedeok-nvidia commented:
Thanks @qixiang-99 for addressing the last-minute reviews as well.  @eopXD could you help with another round of review?

tensorrt-cicd commented:
[PR_Github #15190](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15190/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11472](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11472/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #15206](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15206/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #15349](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15349/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15349](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15349/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11581](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11581/) completed with status: 'FAILURE'

qixiang-99 committed:
revert: revert the change for prepare estimation:
- we shouldn't use max_seq_len as the kv config max_tokens as it doesn't need that much
- and it makes preparation to have OOM more easily especially long seq.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #15464](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15464/) [ run ] triggered by Bot

eopXD reviewed (approved):
LGTM, thank you for the test coverages!

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #16082](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16082/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #12093](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12093/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #16130](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16130/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --disable-fail-fast

eopXD commented:
/bot run --disable-fail-fast
